### PR TITLE
[DNM] proof of concept nullspacing of undertiles. last big maptick optimization?

### PR DESCRIPTION
**Pull request by Kylerace**
## About The Pull Request
this is a proof of concept pr that nullspaces undertile objects to reduce movables being processed by sendmaps.
"proof of concept" meaning that the code is abysmal but technically works in what ive tested to the point that we should be able to test merge this for a while

had this idea for quite a while but was too scared to execute it since it would require some sort of fake location association, so i decided fuck it and to just make a proof of concept to test merge and see what the results are. you may not know it but pipes, wires, and disposal tubes are always physically on the tile theyre placed on but invisible. you may think that this is ok, but SendMaps processes movables in range of clients regardless of whether said clients can see them. and there is quite a lot of undertile objects. 
Screenshot_1210
heres a traffic heavy part of metastation for example, each and every one of these pipes, wires, and disposal tubes are physically on those tiles contributing to the movable count and increasing maptick. movable count is the most important scaling factor of maptick, the looking for movable changes step of the sendmaps profiler usually takes at least 70% of the SendMaps total cost as you can see in these rounds: https://tgstation13.org/parsed-logs/sybil/data/logs/2021/08/27/round-168552/sendmaps.json https://tgstation13.org/parsed-logs/sybil/data/logs/2021/08/27/round-168564/sendmaps.json and https://tgstation13.org/parsed-logs/sybil/data/logs/2021/08/27/round-168526/sendmaps.json where the percentage of total cost that the movable step takes up is 80.6%, 79.88%, and 75.12% respectively

## Why It's Good For The Game
Screenshot_1208
i made a simple proc to tally stats about movables in range for maptick, the stat orderings are bad but in most places i go on station around 40% of the total movable count is invisible objects, 99% of which are undertiles. so if undertile objects are around 40% of all movables on average, and movables take around 70-80% of SendMaps cost and we currently have an average maptick per person of 0.5, then we can expect 0.4 * 0.75 * 0.5 = 0.15 less maptick per person from nullspacing undertile objects bringing us down to ~ 0.35

this is significant (and now that im thinking of it, it can be increased by covering up undertile objects in maint but whatever). at 80 pop thats 12% less internal tick usage than current (40% -> 28%).

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
Oh jesus christ.

You want this reviewed? or straight into a tm

**Comment by Kylerace** [comment]
> 
> 
> Oh jesus christ.
> 
> You want this reviewed? or straight into a tm

i dont want people looking at this code with the intention of it being merged its way too shit for that

**Comment by CRITAWAKETS** [comment]
if someone goes in the vents/scrubbers or gets thrown into disposals, what happens to them?

**Comment by Kylerace** [comment]
> 
> 
> if someone goes in the vents/scrubbers or gets thrown into disposals, what happens to them?

forgot about vent crawling so currently it nullspaces you. for disposals though it just moves you into a holder objects (which already existed in the game) thats located on the turf and that moves you

File: code/datums/elements/undertile.dm [hunk 1]
```diff
@@ -1,6 +1,12 @@
+#define CONTENTS_OF(input_atom) input_atom:nullspaced_contents ? input_atom:nullspaced_contents + input_atom:contents : input_atom:contents
+
+#define CREATE_FAKELOC_ASSOCIATION(movable, atom_loc) movable:real_loc = atom_loc; LAZYADD(atom_loc:nullspaced_contents, movable);
+
+#define GET_ASSOCIATED_LOC(movable) movable:real_loc ? movable:real_loc : movable:loc
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/datums/elements/undertile.dm:5]: cope

**Comment by Kylerace** [comment]
image
thanks to @LemonInTheDark again for the stats here. test is a bit shorter than i would have wanted at first but i forgot to ship it with ventcrawling working so i cut it short. 0.08 decrease in maptick per person in 4 days and the average has started to level off. 

so in conclusion doing this the right way might be worth it, but this only saves 6.4% of the tick at 80 pop so ehhh

**Pull request closed without merging**